docs/src/prototypes/MenuButton/MenuButton.tsx

```diff
@@ -0,0 +1,198 @@
+    }
+
+    if (prevState.menuOpen && !this.state.menuOpen) {
+      document.removeEventListener('click', this.handleDocumentClick)
+
+      switch (this.state.lastKeyCode) {
+        case keyboardKey.Enter:
+        case keyboardKey.Escape:
+          focusButton(this.buttonRef)
+          break
+
+        case keyboardKey.Tab:
+          focusNearest(this.buttonRef, this.state.lastShiftKey ? 'previous' : 'next')
+          break
+      }
+    }
+  }
+
+  handleButtonOverrides = (predefinedProps: ButtonProps) => ({
+    onClick: (e: React.SyntheticEvent, buttonProps: ButtonProps) => {
+      _.invoke(predefinedProps, 'onClick', e, buttonProps)
+      this.setState(prevState => ({
+        lastKeyCode: null,
+        menuOpen: !prevState.menuOpen,
+      }))
+    },
+  })
+
+  handleDocumentClick = (e: MouseEvent) => {
+    const { menuOpen } = this.state
+    const target = e.target as HTMLElement
+    const isInside =
+      _.invoke(this.buttonRef.current, 'contains', target) &&
+      _.invoke(this.menuRef.current, 'contains', target)
+
+    if (menuOpen && !isInside) {
+      this.setState({ lastKeyCode: null, menuOpen: false })
+    }
+  }
+
+  handleKeyDown = (e: React.KeyboardEvent) => {
+    const { menuOpen } = this.state
+
+    const keyCode = keyboardKey.getCode(e)
+    const shouldClose =
+      menuOpen && _.includes([keyboardKey.Enter, keyboardKey.Escape, keyboardKey.Tab], keyCode)
+    const shouldOpen =
+      !menuOpen && _.includes([keyboardKey.ArrowDown, keyboardKey.ArrowUp], keyCode)
+
+    if (shouldClose) {
+      // We should prevent default there otherwise event will be bubbled to a button and will cause menu reopen
+      if (keyCode === keyboardKey.Enter) e.preventDefault()
+
+      this.setState({
+        lastKeyCode: keyCode,
+        lastShiftKey: e.shiftKey,
+        menuOpen: false,
+      })
+    }
+
+    if (shouldOpen) {
+      this.setState({ lastKeyCode: keyCode, menuOpen: true })
+    }
+  }
+
+  handleMenuItemOverrides = accessibilityBehavior =>
+    _.map(_.get(this.props.menu, 'items'), (item: ShorthandValue) =>
+      typeof item === 'object'
+        ? {
+            ...item,
+            ...accessibilityBehavior.menuItem,
+          }
+        : {
+            content: item,
+            key: item,
+            ...accessibilityBehavior.menuItem,
+          },
+    )
+
+  render() {
+    const { accessibility, button, disabled, menu, placement } = this.props
+    const { menuOpen } = this.state
+    const accessibilityBehavior: AccessibilityBehavior = accessibility({
+      ...this.props,
+      ...this.state,
+    })
+
+    return (
+      <MenuButtonWrapper onKeyDown={this.handleKeyDown}>
+        <Manager>
+          <Reference>
+            {({ ref }) => (
+              <Ref
+                innerRef={(buttonNode: HTMLButtonElement) => {
+                  // @ts-ignore
+                  this.buttonRef.current = buttonNode
```

Comment:
actually, quite concerned about us doing that. If some day this will be really implemented as read-only prop (at JS level), we would have real problems with that. I would rather expect us following the interface that is suggested by React types

Reply:
Replaced `buttonRef` with `buttonNode` and `menuRef` with `menuNode`, agree with you there :)

---

Going to introduce some ref utils to solve these situations:
```jsx
<Ref innerRef={mergeRef(this.buttonRef, ref)}>
//
const buttonRef = toRef(buttonNode)
```